### PR TITLE
concretizer: add mode to reuse dependencies only

### PR DESCRIPTION
**Pull request by michaelkuhn**
This adds a new mode for `concretizer:reuse` called `dependencies`, which only reuses dependencies. Currently, `spack install foo` will reuse older versions of `foo`, which might be surprising to users.

I'm not entirely happy with the argument name (`--reuse-deps`) yet and would argue for making this the default reuse mode.

### CODE REVIEW

**Comment by michaelkuhn** [comment]
@alalazo I have fixed the conflicts and rebased to current develop. Any comments about this change?

**Comment by haampie** [comment]
@becker33 can you coordinate your "reuse from local installs/upstreams" vs "reuse from binary cache" with this PR? Just to avoid that we're ending up with `--reuse="locally installed but not the root spec please"`

**Comment by adamjstewart** [comment]
@becker33 you mentioned that this PR would need some work to update for changes in develop since it was created. What changes are required? Would love to get this in soon, as I think this is the most logical default concretization strategy.

**Comment by michaelkuhn** [comment]
I rebased this to current develop to fix some conflicts. I also did some tests and it still seems to work.

**Review by adamjstewart — DISMISSED** [review]
I tested this with a few individual packages and a massive environment and it works **perfectly**, exactly the behavior I would find most intuitive. Great job on this!

**Comment by michaelkuhn** [comment]
@tgamblin @becker33 ping?

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: I think this is redundant, since it includes specs that have a root as a dependency. The solve will rule that out, but potentially we have more hashes than we want as facts. Did you try:
```suggestion
        # If we only want to reuse dependencies, remove the root specs
        if self.reuse == "dependencies":
            reusable_specs = [
                spec for spec in reusable_specs if not any(root in spec for root in specs)
            ]
```
?

Comment (michaelkuhn) [reply lib/spack/spack/solver/asp.py]: I might be reading the code wrong but aren't we hitting the anonymous case in `__contains__` anyway, which simply calls `satisfies`: https://github.com/spack/spack/blob/develop/lib/spack/spack/spec.py#L4001-L4002
I would therefore assume that this behaves identically. I also did some tests and couldn't provoke any differences.

Comment (alalazo) [reply lib/spack/spack/solver/asp.py]: Reusable specs are all concrete specs collected either from the DB or from buildcaches, root specs can't be anonymous. I checked the ASP before posting. Result is the same, we just have a lot of input in the ASP which is useless. 

Try for instance:
1. Install `hdf5+mpi` (fake installs are good for this test)
2. Create an environment with `hdf5+mpi` and zlib as root specs + `reuse:dependencies`
3. `spack solve --show=asp` for the environment

I think you'll see all the specs in there, but `hdf5` and `zlib` There are specs though that reference the non-existing zlib and thus will never be selected.

I think there might be other potential pitfalls we might need to check, e.g. if I can reuse the same package as the root but with different variant in an environment. Going back to the example above do we want to reuse `hdf5~mpi` if it is installed, since the the root says `hdf5+mpi`? That matters if we have a root spec that depend on `hdf5`. I didn't test-drive that case yet.

Comment (michaelkuhn) [reply lib/spack/spack/solver/asp.py]: I got what you meant now, thanks! I confused myself with the "root specs" in the snippet above and actually tested it outside of an environment (which hits the anonymous spec path).

I now did some tests inside an environment and could reproduce your results. I guess the interaction can be a bit surprising if a common dependency is a root spec: I added a new `cmake` version but since `zlib` is a root spec and thus filtered out, `cmake` gets rebuilt even though it could theoretically be reused. I guess there is no way around that, though.

If I add `enzo` (which requires `hdf5~mpi`) in your example above, everything seems to continue working as expected:
```
==> Error: 'hdf5' required multiple values for single-valued variant 'mpi'
    Requested '~mpi' and '+mpi'
```

I also made the change you suggested and pushed a new version.

**Comment by adamjstewart** [comment]
Any updates on this? I find myself needing this behavior again today. My new use case:

I often fix bugs in packages and want to build them to make sure the fix works. I'm faced with two options:

1. `spack install foo`: the package is already installed
2. `spack install --fresh foo`: rebuild the entire world

I would love the feature added in this PR (even more so if it becomes the default):

3. `spack install --reuse-deps foo`: only build foo

**Comment by michaelkuhn** [comment]
> Any updates on this? I find myself needing this behavior again today. My new use case:

I incorporated @alalazo's feedback a while ago, so it should now work better in environments. From my side, this is good to go.

**Comment by michaelkuhn** [comment]
I still run into this *every time* I update a package:
```
$ spack install foo
# already installed
$ spack install foo@x.y.z
```
@alalazo ping :slightly_smiling_face:

**Comment by alalazo** [comment]
I'll try to review this one by the end of the week.

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:2522]: (Comment, not request): If we add more reuse modes, it is worth turning these literal strings into `enums`.

File: lib/spack/spack/test/cmd/common/arguments.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/common/arguments.py]: (minor) Can we parametrize the test?

File: lib/spack/spack/cmd/common/arguments.py
Comment (alalazo) [inline lib/spack/spack/cmd/common/arguments.py:525]: (nitpick) Not sure this is the best way to deal with the command line, but I think this should not be a feature blocker. What I wonder is if:
```
--reuse=<argument>
```
is better and more scalable in general, and if we can deprecate the current `--fresh` argument. @tgamblin @becker33 @haampie Any opinion on this?

**Review by alalazo — DISMISSED** [review]
LGTM. I left a few minor comments. Let's wait a bit to check if people have opinion on the cli flag, otherwise this is ready to be merged.

File: lib/spack/spack/cmd/common/arguments.py
Comment (adamjstewart) [reply lib/spack/spack/cmd/common/arguments.py:525]: Not to say I told you so, but... https://github.com/spack/spack/pull/28468#issuecomment-1015035473

File: lib/spack/spack/test/cmd/common/arguments.py
Comment (michaelkuhn) [reply lib/spack/spack/test/cmd/common/arguments.py]: Done

File: lib/spack/spack/cmd/common/arguments.py
Comment (michaelkuhn) [reply lib/spack/spack/cmd/common/arguments.py:525]: I haven't touched this one since @adamjstewart mentioned on Slack that supporting both `--reuse` and `--reuse=foo` doesn't seem to be possible.

**Review by alalazo — APPROVED** [review]

**Comment by adamjstewart** [comment]
Is there any special handling (in this PR or already in Spack) for handling deprecated versions? Like if one of my packages becomes deprecated, I might not want it to be reused even if it's already installed.

**Comment by alecbcs** [comment]
I don't know of any existing deprecated logic but I too would love this!

**Comment by adamjstewart** [comment]
Either way, it shouldn't hold up this PR since we already reuse everything by default. If there isn't logic to avoid deprecated versions yet, we can add it in a follow-up PR.

**Comment by alecbcs** [comment]
At the moment are there any blockers for this PR? It looks like perhaps we're waiting on final say about `--fresh` vs `--reuse=` options? Or is that just a conversation starter for a follow up?

**Comment by alalazo** [comment]
> At the moment are there any blockers for this PR? It looks like perhaps we're waiting on final say about --fresh vs --reuse= options? Or is that just a conversation starter for a follow up?

I was waiting for that, but since I got no objections to the current state of the PR nor other comments I'll merge this. If objections come up later we can take care of them in a following PR.

**Pull request merged**
